public_html/wp-content/mu-plugins/quickbooks/includes/client.php

```diff
@@ -0,0 +1,382 @@
+<?php
+namespace WordCamp\QuickBooks;
+
+use Exception;
+use QuickBooksOnline\API\DataService\DataService;
+use QuickBooksOnline\API\Core\OAuth\OAuth2\OAuth2AccessToken;
+use QuickBooksOnline\API\Exception\SdkException;
+use QuickBooksOnline\API\Exception\ServiceException;
+use WP_Error;
+
+defined( 'WPINC' ) || die();
+
+/**
+ * Class Client
+ *
+ * This acts as a wrapper around the QBO V3 PHP SDK, so it can catch exceptions and handle errors correctly.
+ *
+ * @package WordCamp\QuickBooks
+ */
+class Client {
+	/**
+	 * The URI that QBO will redirect to after an OAuth user authorization.
+	 *
+	 * This is hardcoded here since it has to be hardcoded in the QBO app.
+	 */
+	const OAUTH_REDIRECT_URI = 'https://wordcamp.org/wp-admin/admin-post.php?action=wordcamp-qbo-oauth&cmd=exchange';
+
+	/**
+	 * @var WP_Error|null
+	 */
+	public $error = null;
+
+	/**
+	 * @var DataService|null
+	 */
+	protected $data_service = null;
+
+	/**
+	 * Client constructor.
+	 */
+	public function __construct() {
+		$this->error = new WP_Error();
+
+		if ( ! class_exists( '\QuickBooksOnline\API\DataService\DataService' ) ) {
+			$this->error->add(
+				'missing_dependency',
+				'The required library <code>QuickBooks V3 PHP SDK</code> is unavailable.'
+			);
+
+			return;
+		}
+
+		$config = $this->get_client_config();
+
+		if ( empty( $config['ClientID'] ) || empty( $config['ClientSecret'] ) ) {
+			$this->error->add(
+				'missing_credentials',
+				'The required credentials for connecting to the QBO API are unavailable.'
+			);
+
+			return;
+		}
+
+		try {
+			$this->data_service = DataService::Configure( $config );
+		} catch ( SdkException $exception ) {
+			$this->add_error_from_exception( $exception );
+		}
+
+		if ( isset( $config['refreshTokenKey'] ) ) {
+			try {
+				/**
+				 * The OAuth access token expires after an hour, so whenever we instantiate with saved token data, we
+				 * need to refresh it with the refresh token to get a current access token. According to the SDK docs,
+				 * the refresh token may change when this happens, so it needs to be re-saved during this process.
```

Comment:
Interesting. So, to make sure I understand this– this code will fetch a new access token whenever the client is initialized, which should ensure that any later call has a valid token. Is that right?

Reply:
Yes, it will ensure that all calls on the current session have a valid access token (unless the session somehow lasts more than an hour).

This might be slightly overkill, but it's simpler than storing the access token expiration timestamp and then writing logic to decide whether there's enough time left on the token or not before refreshing it.